pkg/controllers/expiration/controller.go

```diff
@@ -0,0 +1,123 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package expiration
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha2"
+	"go.uber.org/zap"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"knative.dev/pkg/ptr"
+
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// Controller for the resource
+type Controller struct {
+	kubeClient client.Client
+}
+
+// NewController constructs a controller instance
+func NewController(kubeClient client.Client) *Controller {
+	return &Controller{
+		kubeClient: kubeClient,
+	}
+}
+
+// Reconcile executes an expiration control loop for a node
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	// 1. Get node
+	node := &v1.Node{}
+	if err := c.kubeClient.Get(ctx, req.NamespacedName, node); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// 2. Ignore if node is already deleting
+	if node.DeletionTimestamp != nil {
+		return reconcile.Result{}, nil
+	}
+	// 3. Ignore if provisioner doesn't exist
+	name, ok := node.Labels[v1alpha2.ProvisionerNameLabelKey]
+	if !ok {
+		return reconcile.Result{}, nil
+	}
+	namespace, ok := node.Labels[v1alpha2.ProvisionerNamespaceLabelKey]
+	if !ok {
+		return reconcile.Result{}, nil
+	}
+	provisioner := &v1alpha2.Provisioner{}
+	if err := c.kubeClient.Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, provisioner); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// 4. Ignore if TTLSecondsUntilExpired isn't defined
+	if provisioner.Spec.TTLSecondsUntilExpired == nil {
+		return reconcile.Result{}, nil
+	}
+	// 5. Trigger termination workflow if expired
+	expirationTTL := time.Duration(ptr.Int64Value(provisioner.Spec.TTLSecondsUntilExpired)) * time.Second
+	expirationTime := node.CreationTimestamp.Add(expirationTTL)
+	if time.Now().After(expirationTime) {
+		zap.S().Infof("Triggering termination after %s (+%s) for expired node %s", expirationTTL, time.Since(expirationTime), node.Name)
```

Comment:
What do you think about referencing the name of the Expiration controller here to make it more explicit that it was the expiry controller? It could be confusing not knowing which type of expiration it was, considering we have this concept for utilization.

Reply:
This is the difference between "empty" and "expired". I'd prefer to leave as is, but I do want to do a followup PR (using kn/pkg) which prepends a [Expiration] or [Utilization] to all of the log lines here.